tests/utils_tests/test_functional.py

```diff
@@ -47,43 +49,107 @@ def __bytes__(self):
+            self.assertTrue(callable(inst.other_value))
+
     def test_cached_property(self):
         """
         cached_property caches its value and that it behaves like a property
         """
-        class A:
+        class Class:
 
             @cached_property
             def value(self):
                 """Here is the docstring..."""
                 return 1, object()
 
+            @cached_property
+            def __foo__(self):
+                """Here is the docstring..."""
+                return 1, object()
+
             def other_value(self):
-                return 1
+                """Here is the docstring..."""
+                return 1, object()
 
             other = cached_property(other_value, name='other')
 
-        # docstring should be preserved
-        self.assertEqual(A.value.__doc__, "Here is the docstring...")
+        attrs = ['value', 'other', '__foo__']
+        for attr in attrs:
+            self.assertCachedPropertyWorks(attr, Class)
+
+    @unittest.skipUnless(PY36, '__set_name__ is new in Python 3.6')
+    def test_cached_property_auto_name(self):
+        """
+        cached_property caches its value and that it behaves like a property
+        on mangled methods or no name kwarg set.
+        """
+        class Class:
 
-        a = A()
+            @cached_property
+            def __value(self):
+                """Here is the docstring..."""
+                return 1, object()
 
-        # check that it is cached
-        self.assertEqual(a.value, a.value)
+            def other_value(self):
+                """Here is the docstring..."""
+                return 1, object()
 
-        # check that it returns the right thing
-        self.assertEqual(a.value[0], 1)
+            other = cached_property(other_value)
+            other2 = cached_property(other_value, name='different_name')
 
-        # check that state isn't shared between instances
-        a2 = A()
-        self.assertNotEqual(a.value, a2.value)
+        attrs = ['_Class__value', 'other']
+        for attr in attrs:
+            self.assertCachedPropertyWorks(attr, Class)
 
-        # check that it behaves like a property when there's no instance
-        self.assertIsInstance(A.value, cached_property)
+        inst = Class()
 
         # check that overriding name works
-        self.assertEqual(a.other, 1)
-        self.assertTrue(callable(a.other_value))
+        self.assertEqual(inst.other2[0], 1)
+        self.assertEqual(inst.different_name[0], 1)
+        self.assertNotEqual(inst.other2, inst.other2)
+        self.assertEqual(inst.different_name, inst.different_name)
+        before = inst.different_name
+        during = inst.other2
+        after = inst.different_name
+        self.assertNotEqual(before, after)
+        self.assertNotEqual(before, during)
+        self.assertEqual(during, after)
+
+    @unittest.skipUnless(not PY36, '__set_name__ is new in Python 3.6')
```

Comment:
You'll want to branch of lesser than Python 3.6 here.

Reply:
that's how this works already